validator/src/main/java/io/avaje/validation/core/CoreAdapterBuilder.java

```diff
@@ -73,12 +76,18 @@ <T> ValidationAdapter<T> build(Type type, Object cacheKey) {
   <T> ValidationAdapter<T> buildAnnotation(Class<? extends Annotation> cls, Map<String, Object> attributes) {
     // Ask each factory to create the validation adapter.
     for (final ValidationContext.AnnotationFactory factory : annotationFactories) {
-      final var result = (ValidationAdapter<T>) factory.create(cls, context, attributes);
+      final var result = (ValidationAdapter<T>) factory.create(cls, context, extractGroups(attributes), attributes);
       if (result != null) {
         return result;
       }
     }
     // unknown annotations have noop
     return NOOP;
   }
+
+  static Set<Class<?>> extractGroups(Map<String, Object> attributes) {
+    @SuppressWarnings("unchecked")
+    final var annotationsGroups = (List<Class<?>>) attributes.getOrDefault("groups", List.of());
```

Comment:
At generation time we could put immutable set of class into the attributes. Then we could skip the list -> set conversion here ?

Reply:
I could have done that, but I forgot